documentation/testing_plans.md

```diff
@@ -0,0 +1,783 @@
+  return $results
+}
+```
+
+However, the test would fail for the following plan, since the plan runs the
+`facts` task twice:
+
+```puppet
+plan my_project::my_plan (
+  TargetSpec $servers,
+  TargetSpec $databases
+) {
+  $server_results   = run_task('facts', $servers)
+  $database_results = run_task('facts', $databases)
+  $results          = ResultSet.new($server_results.results + $database_results.results)
+  return $results
+}
+```
+
+Modifiers can also be chained, allowing you to add as much or as little
+specificity to a stub or mock as you want. For example, you can restrict a task
+to run on a specific list of targets with a specific set of parameters:
+
+```ruby
+it 'configures servers' do
+  # Allow the 'my_project::configure' task to be run
+  allow_task('my_project::configure')
+
+  # Assert that the 'my_project::configure' task is run
+  expect_task('my_project::configure').with_targets('servers').with_params({ 'confpath' => '/path/to/conf' })
+
+  # Run the plan
+  run_plan('my_project::configure', 'targets' => ['servers', 'databases'])
+end
+```
+
+📖 **Related information**
+
+- See the [BoltSpec reference](boltspec_reference.md) for a full list of
+  modifiers.
+
+### Matching stubs and mocks
+
+When you run your tests and a plan invokes an executor function, `BoltSpec` will
+match the function invocation to any stubs or mocks created for the test.
+Executor function invocations are matched to any stub or mock that is as
+specific or less specific than the function invocation, but not to any stub or
+mock that is more specific than the function invocation. If the function
+invocation matches multiple stubs or mocks, it will use the last stub or mock
+that matched.
+
+For example, the following plan invokes the `run_task` function to run the
+`configure` task on a list of targets:
+
+And the following plan which runs the `configure` task on a list of targets:
+
+```puppet
+plan configure (
+  TargetSpec $targets
+) {
+  $result = run_task('configure', $targets)
+  return $result
+}
+```
+
+Given the following test which has three `expect_task` mocks:
+
+```ruby
+it 'runs the configure task' do
+  expect_task('configure')
+  expect_task('configure').with_targets(['servers'])
+  expect_task('configure').with_targets(['servers']).with_params('autoupdate' => true)
+
+  run_plan('configure', 'targets' => 'servers')
+end
+```
+
+`BoltSpec` will match the `run_task` invocation to the following mocks:
+
+```ruby
+expect_task('configure')
+expect_task('configure').with_targets(['servers'])
+```
+
+`BoltSpec` does not match the `run_task` invocation to the third mock because it
+uses the `with_params` modifier to add more specificity. Because `BoltSpec`
+found multiple matching mocks, it will use the last one defined in the test:
+`expect_task('configure').with_targets(['servers'])`. The test will then fail,
+as the first mock is not fulfilled, as the plan does not run the `configure`
+task a second time.
```

Comment:
All mocks need their assertions to be fulfilled for a test to pass since a mock is an assertion that something will happen. When `BoltSpec` uses a mock, it increments a counter on the mock to indicate how many times it has been met. Because `BoltSpec` only uses the last matching mock, only the last matching mock will have its counter incremented.

Reply:
Okay cool, I updated this edit and added a sentence above to clarify that.